src/odemis/driver/tfsbc.py

```diff
@@ -0,0 +1,357 @@
+        register_values = transform_coordinates(value, xlower, ylower, xupper, yupper)
+
+        # Read previous value of registers for debugging purpose
+        # Note on duration: a write instruction takes about 14 ms, a read instruction about 20 ms
+        ret = self._read_registers()
+        logging.debug("Register values before writing: %s." % ret)
+
+        logging.debug("Writing register values %s" % register_values)
+        self._write_registers(register_values)
+
+        # Convert back to original coordinates (should be the same as requested shift, possibly
+        # with a small rounding error)
+        value = transform_coordinates_reverse(register_values, xlower, ylower, xupper, yupper)
+        return value
+
+    def _write_registers(self, values):
+        """
+        Write to all four registers. Try to reconnect to device in case connection was lost.
+        :values (list of 4 ints): register values (-x, -y, x, y)
+        """
+        if len(values) != 4:
+            raise ValueError("write_registers received payload of invalid length %s != 4." % len(values))
+
+        # Check if values are in allowed range
+        if not all(0 <= val <= 0xFFFF for val in values):
+            raise ValueError("Register values %s not in range [0, 65535]." % values)
+
+        try:
+            # write all registers together (starting at lower x register (=0x01))
+            rq = self._serial.write_registers(BEAMDEFL_LX, values, unit=SLAVE_UNIT)
+        except IOError:
+            self._reconnect()
+            raise IOError("Failed to write registers of beam control firmware, "
+                          "restarted serial connection.")
+
+    def _read_registers(self):
+        """
+        Read all four registers. Try to reconnect to device in case connection was lost.
+        :return (list of 4 ints): register values (-x, -y, x, y)
+        """
+        try:
+            # write all registers together (starting at lower x register (=0x01))
+            rr = self._serial.read_holding_registers(BEAMDEFL_LX, 4, unit=SLAVE_UNIT)
+            return rr.registers
+        except IOError:
+            self._reconnect()
+            raise IOError("Failed to write registers of beam control firmware, "
+                          "restarted serial connection.")
+
+    def _reconnect(self):
+        """
+        Attempt to reconnect the camera. It will block until this happens.
+        On return, the hardware should be ready to use as before, excepted it
+        still needs the settings to be applied.
+        """
+        num_it = 5
+        self.state._set_value(model.HwError("Beam deflection controller disconnected"), force_write=True)
+        logging.warning("Failed to write registers, trying to reconnect...")
+        for i in range(num_it):
+            try:
+                self._serial.close()
+                self._serial.connect()
+                logging.info("Recovered device.")
```

Comment:
That seems a very optimistic code to *reconnect* and check that the device is answering again...
Please implement and test this function.